src/shader/shaderLightSampling.js

```diff
@@ -1,11 +1,51 @@
 export const shaderLightSampling = /* glsl */`
 
+float getSpotAttenuation( const in float coneCosine, const in float penumbraCosine, const in float angleCosine ) {
+
+	return smoothstep( coneCosine, penumbraCosine, angleCosine );
+
+}
+
+float getDistanceAttenuation( const in float lightDistance, const in float cutoffDistance, const in float decayExponent ) {
+
+		// based upon Frostbite 3 Moving to Physically-based Rendering
+		// page 32, equation 26: E[window1]
+		// https://seblagarde.files.wordpress.com/2015/07/course_notes_moving_frostbite_to_pbr_v32.pdf
+		float distanceFalloff = 1.0 / max( pow( lightDistance, decayExponent ), EPSILON );
+
+		if ( cutoffDistance > 0.0 ) {
+			distanceFalloff *= pow2( saturate( 1.0 - pow4( lightDistance / cutoffDistance ) ) );
+		}
+
+		return distanceFalloff;
+}
+
+// float atan2(float y, float x) {
+
+//     bool s = ( abs( x ) > abs( y ) );
+//     return mix( PI / 2.0 - atan( x, y ), atan( y, x ), s );
+
+// }
+
+float getPhotometricAttenuation(int iesProfile, vec3 posToLight, vec3 lightDir, vec3 u, vec3 v) {
+    float cosTheta = dot(-posToLight, lightDir);
+    float angle = acos(cosTheta) * (1.0 / PI);
+
+	//float y = dot( posToLight, u ); // TODO: use azimuth if any IES profiles require this
+	//float x = dot( posToLight, v );
+	//float a2 = atan2( y, x );
+
+    return texture2D( iesProfiles, vec3( 0.0, angle, iesProfile ) ).r;
+}
```

Comment:
Similar to `readMaterialInfo` can we pass the iesTexture into this function? If possible I'd like to keep these utility functions in other files from reading uniforms directly so they can more easily be reused if needed.

Reply:
Can you elaborate? I might be misunderstanding. Can't we adjust the function signature to the following?

```glsl
float getPhotometricAttenuation( sampler2DArray iesTexture, int iesProfile, vec3 posToLight, vec3 lightDir, vec3 u, vec3 v ) {
```

Reply:
Apologies, I wasn't looking at the code.

> Can't we adjust the function signature to the following? ...

That's how I'd interpreted your initial point.  

We can thread the `iesTexture` through by passing it to `randomLightSample`, in some applications you can apply them to any light type so it isn't inappropriate to pass it to this function.